prov/tcp/src/tcpx_ep.c

```diff
@@ -433,6 +433,70 @@ static void tcpx_ep_tx_rx_queues_release(struct tcpx_ep *ep)
 	fastlock_release(&ep->lock);
 }
 
+static int tcpx_match_ctx(struct slist_entry *item, const void *arg)
+{
+	struct tcpx_xfer_entry *xfer_entry;
+
+	xfer_entry = container_of(item, struct tcpx_xfer_entry, entry);
+
+	return (xfer_entry->context == arg);
+}
+
+static int tcpx_ep_cancel_queue(struct slist *queue, void *context,
+				struct tcpx_cq *tcpx_cq)
+{
+	struct slist_entry *entry;
+	struct tcpx_xfer_entry *xfer_entry;
+
+	entry = slist_remove_first_match(queue, &tcpx_match_ctx, context);
+	if (!entry)
+		return 0;
+
+	xfer_entry = container_of(entry, struct tcpx_xfer_entry, entry);
+
+	tcpx_cq_report_error(&tcpx_cq->util_cq, xfer_entry, FI_ECANCELED);
+	tcpx_xfer_entry_release(tcpx_cq, xfer_entry);
+	return 1;
+}
+
+static ssize_t tcpx_ep_cancel(fid_t fid, void *context)
+{
+	struct tcpx_cq *tcpx_cq;
+	struct tcpx_ep *ep;
+	int ret;
+
+	ep = container_of(fid, struct tcpx_ep,
+			  util_ep.ep_fid.fid);
+
+	fastlock_acquire(&ep->lock);
+
+	tcpx_cq = container_of(ep->util_ep.tx_cq, struct tcpx_cq, util_cq);
+	ret = tcpx_ep_cancel_queue(&ep->tx_queue, context, tcpx_cq);
+	if (ret)
+		goto out;
+
+	ret = tcpx_ep_cancel_queue(&ep->rma_read_queue, context, tcpx_cq);
+	if (ret)
+		goto out;
+
+	ret = tcpx_ep_cancel_queue(&ep->tx_rsp_pend_queue, context, tcpx_cq);
```

Comment:
This queue is for transfers that are waiting for an ACK like response from the peer.  The code assumes that any ACK will find a match (there's an assert for this -- see tcpx_progress.c::367).  By adding cancel, it opens the possibility that an unmatched ACK can occur.  Additional changes are needed to handle this.  This could be added as a separate patch applied prior to this change.

A similar set of changes would be needed for the rma path, but those changes may be more involved.